### PR TITLE
fix: relax OAuth2 app form validation for names and callback URLs

### PR DESCRIPTION
**Pull request by matifali**
## Problem

The OAuth2 application settings form rejects application names with spaces and callbacks using native application schemes. A VS Code extension client registered through DCR exposed this: `VS Code Coder Extension` and `vscode://coder.coder-remote/oauth/callback` were already stored, but the admin form prevented editing them.

## Fix

- Allow application names with spaces and punctuation, limited to 64 UTF-8 bytes on both client and server.
- Allow native callback schemes while rejecting dangerous schemes, unsupported URNs, opaque callbacks such as `localhost:3000`, and empty targets such as `vscode:`. HTTP(S) callbacks still require a host.
- Reuse the public-client registration policy when admins update public clients, so `mailto`, `tel`, `sms`, external HTTP, and fragments remain rejected.
- Trim surrounding whitespace from names and callbacks when submitting, without stripping spaces while typing.
- Update the admin API request validators too, so accepted form values can be saved.

The server remains authoritative for URL parsing. Public-client edits reuse the existing DCR redirect validator using the stored client type; confidential admin-created applications retain their existing HTTP policy. Native host/path callbacks and the supported OOB URN remain accepted.

## Validation

- Vitest component tests assert create/edit submission payloads, whitespace normalization, Unicode byte limits, and rejection of unsafe or malformed callbacks.
- Storybook stories cover accepted native-client values and invalid callback states.
- Go tests cover name boundaries and accepted/rejected callback values through both admin create and update requests.
- `make pre-commit` and targeted tests run before pushing; CI checked on each revision.

<details>
<summary>Implementation decisions</summary>

- Replace the shared slug validator only for OAuth2 application names; leave user, workspace, and template names unchanged.
- Share the backend's 64-byte name limit with the frontend through the generated `OAuth2AppNameMaxBytes` constant.
- Reuse `codersdk.ValidateRedirectURIScheme` for common admin callback validation. Public-client updates additionally call `codersdk.ValidateRedirectURIs`; confidential admin HTTP behavior is preserved.
- Normalize name and callback whitespace in the submitted payload rather than on each keystroke.
- Keep behavior assertions in Vitest and use Storybook for visual states.

</details>

Closes ENG-3342

> 🤖 This PR was created with the help of Coder Agents, and needs a human review.  🧑💻

### CODE REVIEW

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.stories.tsx [hunk 1]
```diff
@@ -154,6 +154,30 @@ export const DeleteDialogOpen: Story = {
 	},
 };
 
+const dcrApp = {
+	...mockApp,
+	name: "VS Code Coder Extension",
+	callback_url: "vscode://coder.coder-remote/oauth/callback",
+};
```
Comment (jeremyruppel) [inline site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.stories.tsx]: probably fine to inline this if it's only used once

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 2]
```diff
@@ -38,25 +33,46 @@ type OAuth2AppFormProps = {
...
 		return false;
 	}
 };
 
 const validationSchema = Yup.object({
-	name: nameValidator("Name"),
+	name: Yup.string()
+		.trim()
+		.required("Please enter a name.")
+		.max(64, "Name cannot be longer than 64 characters."),
```
Comment (jeremyruppel) [inline site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Client/server mismatch on leading/trailing whitespace in the name.

`Yup.string().trim()` in non-strict mode is a *transform*, not an assertion: it trims the value before validating, so `" foo "` passes here. But Formik submits the raw form state (not the cast value), and this PR also removes `onChangeTrimmed` from the Name field. Meanwhile the backend `OAuth2AppNameValid` rejects leading/trailing whitespace (`strings.TrimSpace(str) != str`, covered by the new `NameLeadingSpace` test).

Net effect: a name with surrounding spaces passes the form but the save 400s — the same "form passes, save fails" class this PR sets out to fix. Verified against the repo's Yup:

```
isValid(' foo '): true
cast:  {"name":"foo"}
raw stays: {"name":" foo "}
```

Options to close the gap: `Yup.string().strict().trim()` (surfaces a visible error), a `.test()` asserting `value === value.trim()`, or trim `name` in `onSubmit`. Edge case, but worth aligning with the backend rule.

File: coderd/oauth2provider/provider_test.go [hunk 3]
```diff
@@ -125,6 +119,35 @@ func TestOAuth2ProviderAppValidation(t *testing.T) {
...
+		t.Parallel()
+
+		client := coderdtest.New(t, nil)
+		_ = coderdtest.CreateFirstUser(t, client)
+		ctx := testutil.Context(t, testutil.WaitLong)
+
+		// Values registered through Dynamic Client Registration (RFC 7591),
+		// such as a name with spaces and a custom native-app callback scheme,
+		// must be creatable and editable via the admin OAuth2 app settings.
+		//nolint:gocritic // OAuth2 app management requires owner permission.
```
Comment (jeremyruppel) [inline coderd/oauth2provider/provider_test.go]: This `//nolint:gocritic` looks unnecessary. The custom `testingWithOwnerUser` ruleguard in `scripts/rules.go` only matches the enterprise constructor (`coderdenttest.New`) for method calls; this test uses the AGPL `coderdtest.New(t, nil)`, which no ruleguard pattern matches.

Verified: stripping the `//nolint:gocritic` comments and running `golangci-lint run ./coderd/oauth2provider/...` reports `0 issues`. Safe to drop here (the pre-existing ones in this file are the same cargo-culted pattern, but that's out of scope for this PR).

File: coderd/oauth2provider/provider_test.go [hunk 4]
```diff
@@ -125,6 +119,35 @@ func TestOAuth2ProviderAppValidation(t *testing.T) {
...
+		app, err := client.PostOAuth2ProviderApp(ctx, codersdk.PostOAuth2ProviderAppRequest{
+			Name:        "VS Code Coder Extension",
+			CallbackURL: "vscode://coder.coder-remote/oauth/callback",
+		})
+		require.NoError(t, err)
+		require.Equal(t, "VS Code Coder Extension", app.Name)
+		require.Equal(t, "vscode://coder.coder-remote/oauth/callback", app.CallbackURL)
+
+		//nolint:gocritic // OAuth2 app management requires owner permission.
```
Comment (jeremyruppel) [inline coderd/oauth2provider/provider_test.go]: Same as above — this `//nolint:gocritic` can be dropped; the ruleguard doesn't fire on `coderdtest.New` clients.

**Review by jeremyruppel — COMMENTED** [review]
A couple of review notes on the validation relaxation. Left inline on the specific lines.

_Posted by Coder Agents on behalf of @jeremyruppel._

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.stories.tsx (hunk 1 above)
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.stories.tsx]: Inlined the single-use fixture in 74df9adf3b.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx (hunk 2 above)
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in 74df9adf3b. Trim the name at submission, preserving spaces while typing. Added create/edit payload tests for surrounding whitespace; verified they fail before the fix and pass afterward.

> 🤖 This response was generated by Coder Agents.

File: coderd/oauth2provider/provider_test.go (hunk 3 above)
Comment (matifali) [reply coderd/oauth2provider/provider_test.go]: Removed in 74df9adf3b. Full make pre-commit, including Go lint, passes.

> 🤖 This response was generated by Coder Agents.

File: coderd/oauth2provider/provider_test.go (hunk 4 above)
Comment (matifali) [reply coderd/oauth2provider/provider_test.go]: Removed in 74df9adf3b.

> 🤖 This response was generated by Coder Agents.

**Comment by matifali** [comment]
/coder-agents-review

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 5]
```diff
@@ -81,7 +97,7 @@ export const OAuth2AppForm: FC<OAuth2AppFormProps> = ({
 		validateOnMount: true,
 		onSubmit: async (values) => {
 			didSubmit.current = true;
-			await onSubmit(values);
+			await onSubmit({ ...values, name: values.name.trim() });
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in b37807eb65. Trimmed callback_url at submit, with leading/trailing whitespace payload tests.

> 🤖 This response was generated by Coder Agents.

File: codersdk/name.go [hunk 6]
```diff
@@ -59,6 +59,26 @@ func NameValid(str string) error {
...
 
+// OAuth2AppNameValid returns whether the input string is a valid OAuth2
+// application name. It mirrors the constraints applied to dynamically
+// registered clients (RFC 7591 client_name) so that values registered through
+// DCR remain editable via the admin OAuth2 app settings: a non-empty string of
+// at most 64 characters with no leading or trailing whitespace. Unlike
+// NameValid, it allows spaces and other characters that appear in real client
+// names such as "VS Code Coder Extension".
+func OAuth2AppNameValid(str string) error {
+	if len(str) < 1 {
```
[the remark this answers is not in the document]
Comment (matifali) [reply codersdk/name.go:72]: Addressed in b37807eb65. Retained the backend 64 UTF-8 byte limit, aligned the frontend with TextEncoder, and added multibyte boundary tests on both sides.

> 🤖 This response was generated by Coder Agents.

File: coderd/httpapi/httpapi.go [hunk 7]
```diff
@@ -49,13 +50,54 @@ func init() {
...
+	if err != nil {
+		panic(err)
+	}
+
+	// oauth2_callback_url accepts any redirect URI scheme that DCR accepts
+	// (RFC 7591), including custom native-app schemes such as vscode://, so
+	// that DCR-registered clients stay editable via the admin OAuth2 app
+	// settings. Dangerous schemes (javascript, data, file, ftp) and
+	// unsupported URNs are still rejected.
+	oauth2CallbackURLValidator := func(fl validator.FieldLevel) bool {
```
[the remark this answers is not in the document]
Comment (matifali) [reply coderd/httpapi/httpapi.go:75]: Addressed in b37807eb65. Corrected the validator comment and PR body: this is the admin scheme/HTTP-host policy, not full DCR validation.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 8]
```diff
@@ -38,25 +33,46 @@ type OAuth2AppFormProps = {
 
 const BACK_HREF = "/deployment/oauth2-provider/apps";
 
-const isHttpUrl = (value: string | undefined): boolean => {
+// Reject unsafe callback schemes without restricting native-app schemes.
+// oxlint-disable-next-line eslint/no-script-url -- This blocklist rejects the scheme; it is never used as a navigation target.
+const DANGEROUS_CALLBACK_SCHEMES = ["javascript:", "data:", "file:", "ftp:"];
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx:40]: Addressed in b37807eb65. Documented server authority and cross-referenced the Go validator. Added browser/API cases for opaque localhost, uppercase OOB URN and malformed HTTP inputs; the frontend now rejects hostless HTTP shorthand. Full parser equivalence is intentionally not claimed.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 9]
```diff
@@ -38,25 +33,46 @@ type OAuth2AppFormProps = {
...
+		.required("Please enter a name.")
+		.max(64, "Name cannot be longer than 64 characters."),
 	callback_url: Yup.string()
 		.trim()
 		.required("Please enter a callback URL.")
-		.test("http-url", "Callback URL must be a valid URL.", (value) =>
-			isHttpUrl(value),
+		.test(
+			"valid-callback-url",
+			"Callback URL must be a valid URL and cannot use a dangerous scheme (e.g. javascript, data, file, ftp).",
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in b37807eb65. Replaced the diagnostic with neutral wording requesting a valid callback URL with an allowed scheme; it no longer labels every rejection dangerous.

> 🤖 This response was generated by Coder Agents.

File: coderd/oauth2provider/provider_test.go [hunk 10]
```diff
@@ -55,13 +56,6 @@ func TestOAuth2ProviderAppValidation(t *testing.T) {
...
 				},
 			},
-			{
-				name: "URLLocalhostNoScheme",
-				req: codersdk.PostOAuth2ProviderAppRequest{
-					Name:        "foo",
-					CallbackURL: "localhost:3000",
-				},
-			},
 			{
```
[the remark this answers is not in the document]
Comment (matifali) [reply coderd/oauth2provider/provider_test.go:62]: Addressed in b37807eb65. Added accepted opaque localhost:3000 coverage through both POST and PUT, plus frontend submission coverage.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 11]
```diff
@@ -38,25 +33,46 @@ type OAuth2AppFormProps = {
...
+const DANGEROUS_CALLBACK_SCHEMES = ["javascript:", "data:", "file:", "ftp:"];
+
+const isValidCallbackURL = (value: string | undefined): boolean => {
 	if (!value) {
 		return false;
 	}
 	try {
 		const url = new URL(value);
-		return url.protocol === "http:" || url.protocol === "https:";
+		if (!url.protocol) {
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in b37807eb65. Removed the unreachable protocol guard.

> 🤖 This response was generated by Coder Agents.

File: codersdk/name.go [hunk 12]
```diff
@@ -59,6 +59,26 @@ func NameValid(str string) error {
 	return nil
 }
 
+// OAuth2AppNameValid returns whether the input string is a valid OAuth2
+// application name. It mirrors the constraints applied to dynamically
+// registered clients (RFC 7591 client_name) so that values registered through
+// DCR remain editable via the admin OAuth2 app settings: a non-empty string of
+// at most 64 characters with no leading or trailing whitespace. Unlike
```
[the remark this answers is not in the document]
Comment (matifali) [reply codersdk/name.go]: Addressed in b37807eb65. Corrected the docstring and PR body to describe the actual byte/whitespace rules, without claiming full DCR parity.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.test.tsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+		await user.type(screen.getByLabelText(/^name/i), "test-app");
+		const callbackURL = screen.getByLabelText(/callback url/i);
+		// oxlint-disable-next-line eslint/no-script-url -- Deliberately invalid input exercises callback URL rejection.
+		await user.type(callbackURL, "javascript:alert(1)");
+		await user.click(
+			screen.getByRole("button", { name: /create application/i }),
+		);
+
+		expect(onSubmit).not.toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.test.tsx:110]: Addressed in b37807eb65. Flush pending React updates with async act before negative submit assertions. Expanded rejection coverage.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in b37807eb65. Repository settings allow squash merges only (merge/rebase disabled). The PR title describes the change; no published history rewrite needed.

> 🤖 This response was generated by Coder Agents.

**Comment by matifali** [comment]
Addressed the review findings in b37807eb65. CRF-2 now has positive and negative callback tests through both POST and PUT, asserting 400 callback-field errors. CRF-11 is pre-existing and outside the diff; leaving that duplicate doc-comment cleanup for a follow-up. The repository only permits squash merges, so CRF-12 needs no history rewrite.

Validation: make pre-commit, 23 frontend behavior tests, 12 Storybook tests, and targeted Go validation tests pass.

> 🤖 This response was generated by Coder Agents.

**Comment by matifali** [comment]
/coder-agents-review

> 🤖 This response was generated by Coder Agents.

File: codersdk/name.go [hunk 14]
```diff
@@ -59,6 +59,24 @@ func NameValid(str string) error {
 	return nil
 }
 
+// OAuth2AppNameValid returns whether the input string is a valid OAuth2
+// application name: a non-empty string of at most 64 UTF-8 bytes with no
+// leading or trailing whitespace. Unlike NameValid, it allows spaces and other
+// characters that appear in dynamically registered client names such as
+// "VS Code Coder Extension".
+func OAuth2AppNameValid(str string) error {
```
[the remark this answers is not in the document]
Comment (matifali) [reply codersdk/name.go:71]: Deferring a control/bidi naming policy to shared admin/DCR hardening. DCR already accepts these names, the admin endpoint is owner-gated, and the consent page uses html/template escaping. An admin-only blacklist would make stored DCR clients uneditable again and broadly rejecting Cf would also reject legitimate formatting such as emoji ZWJ. This PR relaxes a slug constraint; it does not define a new Unicode display-security policy.

> 🤖 This response was generated by Coder Agents.

File: coderd/httpapi/httpapi.go [hunk 15]
```diff
@@ -49,13 +50,52 @@ func init() {
...
+		u, err := url.Parse(str)
+		if err != nil {
+			return false
+		}
+		if err := codersdk.ValidateRedirectURIScheme(u); err != nil {
+			return false
+		}
+		// http(s) callback URLs must include a host; custom native-app schemes
+		// (e.g. vscode://) may be opaque.
+		if (u.Scheme == "http" || u.Scheme == "https") && u.Host == "" {
```
[the remark this answers is not in the document]
Comment (matifali) [reply coderd/httpapi/httpapi.go:90]: Keeping the deliberate, documented opaque-scheme contract accepted earlier by the author. A localhost: URI is not HTTP, and exact matching correctly does not equate it with http://localhost. Requiring hosts or an allowlist would reject legitimate opaque native callbacks. Tests now explicitly pin this intentional behavior.

> 🤖 This response was generated by Coder Agents.

File: coderd/oauth2provider/provider_test.go [hunk 16]
```diff
@@ -125,6 +122,88 @@ func TestOAuth2ProviderAppValidation(t *testing.T) {
...
+			{name: "DangerousFileSchemeMixedCase", callbackURL: "FiLe:///tmp/invalid"},
+			{name: "DangerousFTPSchemeMixedCase", callbackURL: "FtP://example.com/invalid"},
+			{name: "UnsupportedURN", callbackURL: "urn:example:invalid"},
+			{name: "UnsupportedURNMixedCase", callbackURL: "URN:example:invalid"},
+		}
+		for _, test := range tests {
+			t.Run(test.name, func(t *testing.T) {
+				t.Parallel()
+
+				client := coderdtest.New(t, nil)
```
[the remark this answers is not in the document]
Comment (matifali) [reply coderd/oauth2provider/provider_test.go]: Addressed in 578f22d9e2: shared the coderd client while retaining unique per-case apps and parallel subtests.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 17]
```diff
@@ -38,25 +33,51 @@ type OAuth2AppFormProps = {
...
-		return url.protocol === "http:" || url.protocol === "https:";
+		if (url.protocol === "urn:") {
+			return url.href === "urn:ietf:wg:oauth:2.0:oob";
+		}
+		if (DANGEROUS_CALLBACK_SCHEMES.includes(url.protocol.toLowerCase())) {
+			return false;
+		}
+		if (
+			(url.protocol === "http:" || url.protocol === "https:") &&
+			(!url.host || !/^https?:\/\/[^/\\\s]/i.test(value))
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in 578f22d9e2: removed the redundant host guard.

> 🤖 This response was generated by Coder Agents.

File: coderd/httpapi/httpapi.go [hunk 18]
```diff
@@ -49,13 +50,52 @@ func init() {
...
+			return false
+		}
+		u, err := url.Parse(str)
+		if err != nil {
+			return false
+		}
+		if err := codersdk.ValidateRedirectURIScheme(u); err != nil {
+			return false
+		}
+		// http(s) callback URLs must include a host; custom native-app schemes
```
[the remark this answers is not in the document]
Comment (matifali) [reply coderd/httpapi/httpapi.go]: Addressed in 578f22d9e2: removed the duplicate comment.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPageView.stories.tsx [hunk 19]
```diff
@@ -80,20 +80,29 @@ export const WithValidationError: Story = {
 	},
 };
 
-export const InvalidName: Story = {
+export const InvalidCallbackURL: Story = {
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPageView.stories.tsx:83]: Keeping play functions assertion-free per the mandatory FE1 contract. Vitest covers rejection and Pixel covers the rendered state; the review verified the negative tests catch mutations. No change warranted.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 20]
```diff
@@ -38,25 +33,51 @@ type OAuth2AppFormProps = {
...
+			(value) => new TextEncoder().encode(value).length <= 64,
+		),
 	callback_url: Yup.string()
 		.trim()
 		.required("Please enter a callback URL.")
-		.test("http-url", "Callback URL must be a valid URL.", (value) =>
-			isHttpUrl(value),
+		.test(
+			"valid-callback-url",
+			"Please enter a valid callback URL with an allowed scheme.",
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in 578f22d9e2: generic invalid-URL wording no longer suggests the scheme is necessarily at fault.

> 🤖 This response was generated by Coder Agents.

**Comment by matifali** [comment]
/coder-agents-review

Please review the latest fixes and explicit dispositions on CRF-14/15/16. Keeping FE1 assertion-free stories, the documented opaque callback contract, and deferring shared Unicode display-policy changes rather than adding an admin-only blacklist. All targeted tests and make pre-commit pass.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.test.tsx [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+			expect(onSubmit).toHaveBeenCalledWith({
+				name: "OAuth App",
+				callback_url: callback.trim(),
+				icon: "",
+			}),
+		);
+	});
+
+	it.each([
+		{ name: "é".repeat(32), valid: true },
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.test.tsx:141]: Addressed in cfce090183. Targeted tests and make pre-commit pass.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 22]
```diff
@@ -38,25 +33,49 @@ type OAuth2AppFormProps = {
...
 	if (!value) {
 		return false;
 	}
 	try {
 		const url = new URL(value);
-		return url.protocol === "http:" || url.protocol === "https:";
+		if (url.protocol === "urn:") {
+			return url.href === "urn:ietf:wg:oauth:2.0:oob";
+		}
+		if (DANGEROUS_CALLBACK_SCHEMES.includes(url.protocol.toLowerCase())) {
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx]: Addressed in cfce090183. Targeted tests and make pre-commit pass.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPageView.stories.tsx [hunk 23]
```diff
@@ -80,20 +80,29 @@ export const WithValidationError: Story = {
...
+		// oxlint-disable-next-line eslint/no-script-url -- Deliberately invalid input exercises callback URL rejection.
+		await userEvent.type(callbackInput, "javascript:alert(1)");
+		await userEvent.tab();
+	},
+};
+
+export const AcceptsDynamicallyRegisteredValues: Story = {
+	play: async ({ canvasElement }) => {
+		const canvas = within(canvasElement);
+		await userEvent.type(
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/CreateOAuth2AppPageView.stories.tsx:97]: Addressed in cfce090183. Targeted tests and make pre-commit pass.

> 🤖 This response was generated by Coder Agents.

File: site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx [hunk 24]
```diff
@@ -81,7 +100,11 @@ export const OAuth2AppForm: FC<OAuth2AppFormProps> = ({
...
 			didSubmit.current = true;
-			await onSubmit(values);
+			await onSubmit({
+				...values,
+				name: values.name.trim(),
+				callback_url: values.callback_url.trim(),
+			});
 		},
 	});
 	const getFieldHelpers = getFormHelpers(form, error);
```
[the remark this answers is not in the document]
Comment (matifali) [reply site/src/pages/DeploymentSettingsPage/OAuth2AppsSettingsPage/OAuth2AppForm.tsx:136]: Addressed in cfce090183. Targeted tests and make pre-commit pass.

> 🤖 This response was generated by Coder Agents.

**Comment by matifali** [comment]
/coder-agents-review

CRF-21 through CRF-24 are addressed in cfce090183. CRF-15 and CRF-16 remain open for human disposition as requested by the latest review; no policy change or acceptance is being inferred from silence. Validation: 24 component cases, 12 Storybook tests and make pre-commit pass.

> 🤖 This response was generated by Coder Agents.

File: codersdk/name.go [hunk 25]
```diff
@@ -59,6 +59,24 @@ func NameValid(str string) error {
...
+	if len(str) < 1 {
+		return xerrors.New("must be >= 1 character")
+	}
+	if len(str) > 64 {
```
Comment (BobbyHo) [inline codersdk/name.go]: NIT: could we create a constant for 64?

File: coderd/httpapi/httpapi.go [hunk 26]
```diff
@@ -49,13 +50,50 @@ func init() {
...
+		u, err := url.Parse(str)
+		if err != nil {
+			return false
+		}
+		if err := codersdk.ValidateRedirectURIScheme(u); err != nil {
+			return false
+		}
+		if (u.Scheme == "http" || u.Scheme == "https") && u.Host == "" {
+			return false
+		}
```
Comment (BobbyHo) [inline coderd/httpapi/httpapi.go:92]: **This now accepts `mailto:`, `tel:` and `sms:` callback URLs, which the docs say public clients can't use.**

The new validator only checks that the scheme is safe and that `http(s)` URLs have a host, so addresses like `mailto:a@b` or `tel:+1…` now pass. The old `http_url` tag rejected them. Because `UpdateApp` keeps the app's existing `ClientType`, an admin can set one of these on a public DCR client through `PUT /oauth2-provider/apps/{app}`.

We block these for public clients at registration on purpose. They open a mail client or dialer instead of returning to the app, so the app can never finish authorizing. Our docs also state this as a guarantee in the *Security Considerations* section of `docs/admin/integrations/oauth2-provider.md`, and there's a troubleshooting entry for the error. After this change, the rule still holds at registration but not here.

I think we should do one of two things:

1. Validate in the handler, where we know the client type. Something like `validateRedirectURIs([]string{req.CallbackURL}, clientType)` would work (it needs exporting). This would also pick up the fragment and http-host rules for free.
2. Update the docs to say the rule only applies at registration.

I'd go with the first option. To be clear, this isn't a security hole. The result is an app that saves fine but can never connect. But it does mean the product stops enforcing something the docs promise.

File: coderd/oauth2provider/provider_test.go [hunk 27]
```diff
@@ -125,6 +122,89 @@ func TestOAuth2ProviderAppValidation(t *testing.T) {
...
+
+		client := coderdtest.New(t, nil)
+		_ = coderdtest.CreateFirstUser(t, client)
+
+		tests := []struct {
+			name        string
+			callbackURL string
+			valid       bool
+		}{
+			{name: "OpaqueLocalhost", callbackURL: "localhost:3000", valid: true},
```
Comment (BobbyHo) [inline coderd/oauth2provider/provider_test.go]: **I'd keep rejecting `localhost:3000` rather than locking in acceptance here.**

This case used to be `URLLocalhostNoScheme` and asserted the opposite. The most likely way someone types `localhost:3000` is by forgetting the `http://` on `http://localhost:3000`, which is an easy mistake when setting up a local tool. Before this change Coder caught it. Now the app saves fine, but nothing can ever redirect to it: the parser reads `localhost` as a scheme, like `vscode`, not as a hostname. The loopback port exception doesn't help either, since it only applies when the registered scheme is `http`. So the developer clicks Allow, the code goes nowhere, and there's no error anywhere to point at the callback URL.

The same thing applies to `vscode:` and `a:`, which also pass now. That's a scheme with nothing after it.

I don't think the ENG-3342 use case needs any of these. Two options:

1. Reject a non-`http(s)` scheme when both the host and the opaque part are empty. That catches `vscode:` and `a:`. Treating a single-word scheme followed only by digits as a typo would catch `localhost:3000` too.
2. Keep accepting it, but add a comment here explaining why a callback nothing can redirect to is a valid value, so the next person doesn't flip this test back.

I'd go with the first. The PR description says this is intentional, so I'm raising it as a behavior question rather than a bug. The old test was guarding a real mistake and I don't see what accepting it buys us.

File: codersdk/name.go (hunk 25 above)
Comment (matifali) [reply codersdk/name.go]: Addressed in 8c4a0ede6c. Added OAuth2AppNameMaxBytes and use its generated TypeScript export, so both sides share the constant.

> 🤖 This response was generated by Coder Agents.

File: coderd/httpapi/httpapi.go (hunk 26 above)
Comment (matifali) [reply coderd/httpapi/httpapi.go:92]: Addressed in 8c4a0ede6c. Public-client admin updates now call the existing redirect URI validator using the stored client type, returning callback_url validation errors. Added API and frontend coverage for mailto/tel/sms (including non-opaque forms), external HTTP and fragments. Confidential admin HTTP behavior is unchanged.

> 🤖 This response was generated by Coder Agents.

File: coderd/oauth2provider/provider_test.go (hunk 27 above)
Comment (matifali) [reply coderd/oauth2provider/provider_test.go]: Addressed in 8c4a0ede6c. Reject opaque callbacks (except the supported OOB URN) and empty targets, including localhost:3000, vscode:, a: and vscode://. Native host/path callbacks remain accepted. Updated frontend/API tests and removed the acceptance claim from the PR body.

> 🤖 This response was generated by Coder Agents.

File: coderd/oauth2provider/apps.go [hunk 28]
```diff
@@ -143,6 +143,18 @@ func UpdateApp(db database.Store, accessURL *url.URL, auditor *audit.Auditor, lo
 		if !httpapi.Read(ctx, rw, r, &req) {
 			return
 		}
+		if app.IsPublic() {
```
Comment (BobbyHo) [inline coderd/oauth2provider/apps.go:146]: NOTE: OAuth apps registered before [PLAT-445](https://linear.app/codercom/issue/PLAT-445/oauth2-support-public-client) may be native/public apps that were issued a secret, so they’re still treated as confidential clients. This logic only applies to newly registered public clients (e.g. via DCR); legacy apps are intentionally unaffected.

**Review by BobbyHo — APPROVED** [review]
The backend changes (Golang) changes lgtm. Thank you for addressing my comments.
